dali/kernels/imgproc/jpeg/chroma_subsample_gpu.cuh

```diff
@@ -0,0 +1,241 @@
+    avg_rgb = avg2(rgb[0], rgb[1]);
+  }
+
+  vec<2, T> cbcr = rgb_to_cb_cr<T>(avg_rgb);
+  cb(chroma_x, chroma_y) = cbcr.x;
+  cr(chroma_x, chroma_y) = cbcr.y;
+}
+
+template <int N, typename T>
+__inline__ __device__
+void write_vec(T* ptr, vec<N, T> v) {
+  #pragma unroll
+  for (int i = 0; i < N; i++)
+    *ptr++ = v[i];
+}
+
+template <bool horz_subsample, bool vert_subsample, typename T>
+__inline__ __device__
+void ycbcr_to_rgb_chroma_subsample(ivec2 blk_offset, ivec2 offset,
+                                   const Surface2D<uint8_t>& out,
+                                   const Surface2D<T>& luma,
+                                   const Surface2D<T>& cb,
+                                   const Surface2D<T>& cr) {
+  int chroma_y = blk_offset.y;
+  int chroma_x = blk_offset.x;
+  int luma_y = chroma_y << vert_subsample;
+  int luma_x = chroma_x << horz_subsample;
+  int y = offset.y;
+  int x = offset.x;
+  T cb_val = cb(chroma_x, chroma_y);
+  T cr_val = cr(chroma_x, chroma_y);
+
+  write_vec(&out(x, y),
+            ycbcr_to_rgb(vec<3, T>(luma(luma_x, luma_y), cb_val, cr_val)));
+  if (horz_subsample && vert_subsample) {
+    write_vec(&out(x + 1, y),
+              ycbcr_to_rgb(vec<3, T>(luma(luma_x + 1, luma_y), cb_val, cr_val)));
+    write_vec(&out(x, y + 1),
+              ycbcr_to_rgb(vec<3, T>(luma(luma_x, luma_y + 1), cb_val, cr_val)));
+    write_vec(&out(x + 1, y + 1),
+              ycbcr_to_rgb(vec<3, T>(luma(luma_x + 1, luma_y + 1), cb_val, cr_val)));
+  } else if (horz_subsample) {
+    write_vec(&out(x + 1, y),
+              ycbcr_to_rgb(vec<3, T>(luma(luma_x + 1, luma_y), cb_val, cr_val)));
+  } else if (vert_subsample) {
+    write_vec(&out(x, y + 1),
+              ycbcr_to_rgb(vec<3, T>(luma(luma_x, luma_y + 1), cb_val, cr_val)));
+  }
+}
+
+
+template <bool horz_subsample, bool vert_subsample>
+__global__ void ChromaSubsampleDistortion(const SampleDesc *samples,
+                                          const kernels::BlockDesc<2> *blocks) {
+  using T = uint8_t;
+  constexpr int luma_blk_h = 8 << vert_subsample;
+  constexpr int luma_blk_w = 8 << horz_subsample;
+  constexpr ivec<2> luma_blk_sz{luma_blk_w, luma_blk_h};
+  constexpr i64vec<2> luma_blk_strides{1, luma_blk_w};
+  constexpr ivec<2> chroma_blk_sz{8, 8};
+  constexpr i64vec<2> chroma_blk_strides{1, 8};
+
+  const auto &block = blocks[blockIdx.x];
+  const auto &sample = samples[block.sample_idx];
+
+  int y_start = threadIdx.y + block.start.y;
+  int x_start = threadIdx.x + block.start.x;
+  if (y_start >= block.end.y || x_start >= block.end.x) {
+    return;
+  }
+
+  // Assuming CUDA block has:
+  // - width 32, leads to 4 horizontal blocks of 8
+  // - height 8, so a single block 8x8 fits vertically
+  __shared__ T luma_blk[4][luma_blk_h][luma_blk_w];
+  __shared__ T cb_blk[4][8][8];
+  __shared__ T cr_blk[4][8][8];
+
+  int blk_idx = threadIdx.x / 8;
+  int local_x = threadIdx.x % 8;
+  int local_y = threadIdx.y;
```

Comment:
I think that having 8x8 blocks for both luma and chroma may simplify the DCT & quantization steps.

```suggestion
  __shared__ float Cb[2][4][8][9];  // yes, 9 - to reduce bank conflicts!
  __shared__ float Cr[2][4][8][9]];
  __shared__ float Y[2 << vert_subsample][4 << horz_subsample][8][9]];
  
  int cx = threadIdx.x;
  int cy = threadIdx.y;
  chroma_x = cx & 7;
  chroma_y = cy & 7;
  chroma_bx = cx >> 3;
  chroma_by = cy >> 3;
  
  int lx = threadIdx.x << horz_subsample;
  int ly = threadIdx.y << vert_subsample;
  luma_x = lx & 7;
  luma_y = ly >> 3;
  luma_bx = lx >> 3;
  luma_by = ly >> 3;
```
With 16x32 block size we'd have
Cr 2*4*8*9 * 4 bytes = 2304 B
Cb 2*4*8*9 * 4 bytes = 2304 B
Y 4*8*8*9 * 4 bytes = 9216 B
total 13824 B - well within acceptable limits, we still have quite a lot of shared mem for DCT step